### PR TITLE
Refactor EventFactory into static methods.

### PR DESCRIPTION
**Pull request by mikecdavis**
## Summary
- Remove EventFactory from Optimizely constructor.
- Remove ClientVersion and ClientEngine from OptimizelyBuilder
- Adds ClientEngineInfo for holding global engine value.
- Defines new UserEvent, ConversionEvent and ImpressionEvent models
- Redefines EventFactory to accept UserEvents via static methods
- Defines UserEventFactory to create UserEvents.

This change is in preparation for batching event payloads so we first want to extract separate canonical representations of conversions and impressions disjoint from the EventBatch model used for sending to the logging backend.

This change is in preparation for an `EventProcessor` abstraction. https://github.com/optimizely/java-sdk/pull/310 is blocked on this change for a cleaner implementation.

### CODE REVIEW

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventFactory.java [hunk 1]
```diff
@@ -16,139 +16,134 @@
...
-        Event conversionEvent = new Event.Builder()
-            .setTimestamp(System.currentTimeMillis())
-            .setUuid(UUID.randomUUID().toString())
-            .setEntityId(eventType.getId())
-            .setKey(eventType.getKey())
-            .setRevenue(EventTagUtils.getRevenueValue(eventTags))
-            .setTags(eventTags)
-            .setType(eventType.getKey())
-            .setValue(EventTagUtils.getNumericValue(eventTags))
+    public static Visitor createConversionEvent(UserEvent userEvent) {
```
Comment (loganlinn) [inline core-api/src/main/java/com/optimizely/ab/event/internal/EventFactory.java]: this should be `private`, no?

File: core-api/src/main/java/com/optimizely/ab/event/internal/ClientEngineInfo.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * ClientEngineInfo is a utility to globally get and set the ClientEngine used in
+ * event tracking. The ClientEngine defaults to JAVA_SDK but can be overridden at
+ * runtime.
+ */
+public class ClientEngineInfo {
+    private static final Logger logger = LoggerFactory.getLogger(ClientEngineInfo.class);
+
+    public static final EventBatch.ClientEngine DEFAULT = EventBatch.ClientEngine.JAVA_SDK;
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/ClientEngineInfo.java]: Yeah I think you can do that for android in general, but then Mobile vs TV is a little more tricky and we do that already [here](https://github.com/optimizely/android-sdk/blob/master/android-sdk/src/main/java/com/optimizely/ab/android/sdk/OptimizelyClientEngine.java). This is a step into making it less clunky and having to pass that info everywhere.

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java [hunk 3]
```diff
@@ -309,22 +301,18 @@ public void track(@Nonnull String eventName,
...
+        UserEvent event = UserEventFactory.createConversionEvent(
             projectConfig,
             userId,
             eventType.getId(),
             eventType.getKey(),
             copiedAttributes,
             eventTags);
 
+        // create the conversion event request parameters, then dispatch
+        LogEvent conversionEvent = EventFactory.createLogEvent(Collections.singletonList(event));
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: Ha yeah, I didn't go back and update all these :)

File: core-api/src/main/java/com/optimizely/ab/event/internal/ConversionEvent.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        public Builder withValue(Number value) {
+            this.value = value;
+            return this;
+        }
+
+        public Builder withTags(Map<String, ?> tags) {
+            this.tags = tags;
+            return this;
+        }
+
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/ConversionEvent.java]: This would probably just be used in testing, but I'll review.

**Comment by mikecdavis** [comment]
After SDK CoP and discussion with @loganlinn I'm going to re-work the `UserEvent` model to remove the "union" behavior it's serving today and instead define a `UserContext` encapsulating the top-level, shared attributes and have a `ConversionEvent` and `ImpressionEvent` take a `UserContext` in their respective builders.

File: core-api/src/main/java/com/optimizely/ab/event/internal/EventFactory.java [hunk 5]
```diff
@@ -16,139 +16,134 @@
...
 import java.util.Map;
-import java.util.UUID;
 
 import static com.optimizely.ab.internal.AttributesUtil.isValidNumber;
 
 public class EventFactory {
     private static final Logger logger = LoggerFactory.getLogger(EventFactory.class);
-    static final String EVENT_ENDPOINT = "https://logx.optimizely.com/v1/events";  // Should be part of the datafile
-    static final String ACTIVATE_EVENT_KEY = "campaign_activated";
+    public static final String EVENT_ENDPOINT = "https://logx.optimizely.com/v1/events";  // Should be part of the datafile
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/EventFactory.java]: This is a valid point, but I want to address it when that feature is supported.

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java [hunk 6]
```diff
@@ -309,22 +301,18 @@ public void track(@Nonnull String eventName,
...
         }
 
-        // create the conversion event request parameters, then dispatch
-        LogEvent conversionEvent = eventFactory.createConversionEvent(
+        UserEvent event = UserEventFactory.createConversionEvent(
             projectConfig,
             userId,
             eventType.getId(),
             eventType.getKey(),
             copiedAttributes,
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: Those attributes get passed around to DecisionService and to Notifications. I think a refactor is a good idea, but will do so in a follow-up PR.

File: core-api/src/main/java/com/optimizely/ab/event/internal/ConversionEvent.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    private final Map<String, ?> tags;
+
+    @VisibleForTesting
+    ConversionEvent() {
+        this(null, null, null, null, null);
+    }
+
+    private ConversionEvent(String eventId, String eventKey, Number revenue, Number value, Map<String, ?> tags) {
+        this.eventId = eventId;
+        this.eventKey = eventKey;
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/ConversionEvent.java]: This is enforced in the `Optimizely#track` call.

File: core-api/src/main/java/com/optimizely/ab/event/internal/ConversionEvent.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        this.revenue = revenue;
+        this.value = value;
+        this.tags = tags;
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/ConversionEvent.java]: This is enforced in the `Optimizely#track` call.

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java [hunk 9]
```diff
@@ -230,36 +225,33 @@ private void sendImpression(@Nonnull ProjectConfig projectConfig,
...
                                 @Nonnull Variation variation) {
-        if (experiment.isRunning()) {
-            LogEvent impressionEvent = eventFactory.createImpressionEvent(
-                projectConfig,
-                experiment,
-                variation,
-                userId,
-                filteredAttributes);
-            logger.info("Activating user \"{}\" in experiment \"{}\".", userId, experiment.getKey());
+        if (!experiment.isRunning()) {
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: Not sure I see the difference? I just flipped the condition that was already there.

[the remark this answers is not in the document]
Comment (aliabbasrizvi) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: Apologies, I misinterpreted this method as part of `activate` where we have already looked up status previously. 
The original comment can be ignored as incorrect.

File: core-api/src/main/java/com/optimizely/ab/event/internal/UserContext.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+package com.optimizely.ab.event.internal;
+
+import com.optimizely.ab.config.ProjectConfig;
+
+import java.util.Map;
+
+/**
+ * UserContext stores the user and project context timestamp and global id.
+ *
+ * Alt name EventContext?
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/UserContext.java]: I might just get rid of this entirely and defer to the interface.

[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/UserContext.java]: eh, I can split this out into `UserContext` and `EventContext`

File: core-api/src/main/java/com/optimizely/ab/event/internal/ClientEngineInfo.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    private static EventBatch.ClientEngine clientEngine = DEFAULT;
+
+    public static void setClientEngine(EventBatch.ClientEngine clientEngine) {
+        if (clientEngine == null) {
+            logger.warn("ClientEngine cannot be null, defaulting to {}", ClientEngineInfo.clientEngine.getClientEngineValue());
+            return;
```
Comment (aliabbasrizvi) [inline core-api/src/main/java/com/optimizely/ab/event/internal/ClientEngineInfo.java]: Where is the default set?

Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/event/internal/ClientEngineInfo.java]: Default is set above. Initial default is `JAVA_SDK`, but can get updated at any time to another valid option.

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java [hunk 12]
```diff
@@ -1047,13 +1032,16 @@ public Builder withUserProfileService(UserProfileService userProfileService) {
             return this;
         }
 
-        public Builder withClientEngine(ClientEngine clientEngine) {
-            this.clientEngine = clientEngine;
+        @Deprecated
+        public Builder withClientEngine(EventBatch.ClientEngine clientEngine) {
+            logger.info("Deprecated. Setting ClientEngine is now set via ClientEngineInfo.");
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: Updated to "Deprecated. In the future, set ClientEngine via ClientEngineInfo#setClientEngine."

File: core-api/src/main/java/com/optimizely/ab/Optimizely.java [hunk 13]
```diff
@@ -230,36 +225,33 @@ private void sendImpression(@Nonnull ProjectConfig projectConfig,
...
-                userId,
-                filteredAttributes);
-            logger.info("Activating user \"{}\" in experiment \"{}\".", userId, experiment.getKey());
+        if (!experiment.isRunning()) {
+            logger.info("Experiment has \"Launched\" status so not dispatching event during activation.");
+            return;
+        }
 
-            if (logger.isDebugEnabled()) {
-                logger.debug(
```
[the remark this answers is not in the document]
Comment (mikecdavis) [reply core-api/src/main/java/com/optimizely/ab/Optimizely.java]: That's a good idea i can move it there.

File: core-api/src/main/java/com/optimizely/ab/event/internal/UserContext.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
```
Comment (aliabbasrizvi) [inline core-api/src/main/java/com/optimizely/ab/event/internal/UserContext.java]: nit. This comment may be out of date now?

**Pull request merged**